test/prop_statem.erl

```diff
@@ -155,6 +158,55 @@ did_close_post(_S, _Args, Res) ->
   ?assertEqual(ok, Res),
   true.
 
+%%------------------------------------------------------------------------------
+%% Shutdown
+%%------------------------------------------------------------------------------
+shutdown() ->
+  erlang_ls_client:shutdown().
+
+shutdown_args(_S) ->
+  [].
+
+shutdown_pre(#{connected := Connected} = _S) ->
+  Connected.
+
+shutdown_next(S, _R, _Args) ->
+  S#{shutdown => true}.
+
+shutdown_post(#{shutdown := true}, _Args, Res) ->
+  assert_shutdown_error(Res),
+  true;
+shutdown_post(_S, _Args, Res) ->
+  ?assertMatch(#{result := null}, Res),
+  true.
+
+%%------------------------------------------------------------------------------
+%% Shutdown
+%%------------------------------------------------------------------------------
+exit() ->
+  erlang_ls_client:exit().
+
+exit_args(_S) ->
+  [].
+
+exit_pre(#{connected := Connected} = _S) ->
+  Connected.
+
+exit_next(S, _R, _Args) ->
+  %% We disconnect to simulate the server goes down
+  catch disconnect(),
+  S#{shutdown => false, connected => false}.
+
+exit_post(S, _Args, Res) ->
+  ExpectedExitCode = case maps:get(shutdown, S, false) of
+                       true  -> 0;
+                       false -> 1
+                     end,
+  wait_for(halt_called, 1000),
+  ?assert(meck:called(erlang_ls_utils, halt, [ExpectedExitCode])),
```

Comment:
I see now why you wanted a wrapper around the halt function.

Reply:
Testing the usage of `erlang:exit/1` always ends up being a little convoluted.